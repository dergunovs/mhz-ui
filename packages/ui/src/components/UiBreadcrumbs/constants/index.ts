export const LINKS = [
  { url: '/', title: 'Main' },
  { url: '/category', title: 'Categories' },
  { url: '/cpu', title: 'CPU' },
];

export const DEFAULT_COLOR = 'default';
export const DEFAULT_ARIA_LABEL = 'Breadcrumb';
