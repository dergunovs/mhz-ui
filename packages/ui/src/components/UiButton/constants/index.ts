export const DEFAULT_LAYOUT = 'primary';
export const DEFAULT_TYPE = 'button';
export const DEFAULT_ICON = undefined;

export const DEFAULT_SLOT = 'Button label';
export const DEFAULT_ARIA_LABEL = 'Button';
