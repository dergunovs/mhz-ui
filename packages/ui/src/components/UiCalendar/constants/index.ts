import { ICalendarEvent } from '../interface';

export const EVENTS: ICalendarEvent<object>[] = [
  {
    id: 'id of event',
    start: new Date(),
    end: new Date(),
    title: '+',
    content: [{ id: 1, text: 'Text' }],
    color: 'linear-gradient(135deg, rgb(168, 60, 255) 36%, rgb(235, 0, 123) 69%, rgb(251, 185, 0) 100%)',
  },
  {
    id: 'id of event',
    start: new Date(),
    end: new Date(),
    title: '+',
    content: [{ id: 2, text: 'Text 2' }],
    color: 'linear-gradient(135deg, rgb(168, 60, 255) 41%, rgb(235, 0, 123) 71%, rgb(0, 134, 255) 100%)',
  },
];
