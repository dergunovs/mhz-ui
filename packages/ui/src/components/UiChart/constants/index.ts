export const LABELS = [
  '07.07 - 13.07',
  '14.07 - 20.07',
  '21.07 - 27.07',
  '28.07 - 03.08',
  '04.08 - 10.08',
  '11.08 - 17.08',
  '18.08 - 24.08',
];

export const DATASETS = [
  {
    data: [28, 16, 12, 4, 22, 16, 14],
    label: 'Грудь',
    borderColor: '#00CED1',
    backgroundColor: '#00CED1',
  },
  {
    data: [4, 6, 0, 2, 2, 0, 6],
    label: 'Ноги',
    borderColor: '#000080',
    backgroundColor: '#000080',
  },
  {
    data: [22, 18, 8, 10, 25, 18, 16],
    label: 'Плечи',
    borderColor: '#C41E3A',
    backgroundColor: '#C41E3A',
  },
  {
    data: [0, 6, 0, 6, 5, 4, 2],
    label: 'Пресс',
    borderColor: '#32CD32',
    backgroundColor: '#32CD32',
  },
  {
    data: [32, 24, 12, 10, 33, 24, 20],
    label: 'Руки',
    borderColor: '#DA70D6',
    backgroundColor: '#DA70D6',
  },
  {
    data: [16, 18, 8, 12, 26, 22, 20],
    label: 'Спина',
    borderColor: '#FF8C00',
    backgroundColor: '#FF8C00',
  },
];

export const TYPE = 'Line';
export const TITLE = 'Chart title';
