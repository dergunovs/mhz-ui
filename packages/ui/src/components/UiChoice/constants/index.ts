export const options = [1, 2, 3, 4, 5];
export const modelValue = options[2];
export const title = 'Заголовок';
export const isTall = true;
export const isInput = true;
