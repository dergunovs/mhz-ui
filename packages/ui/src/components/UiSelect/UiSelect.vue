<template>
  <div :class="$style.container" ref="containerElement">
    <div :class="$style.input">
      <UiInput
        :modelValue="displayValue"
        :disabled="props.isDisabled"
        @toggle="toggleOptions"
        mode="select"
        :placeholder="MESSAGES[props.lang].choose"
        :appendIcon="isShowOptions ? IconOpened : IconClosed"
        data-test="ui-select-input"
      />

      <div v-if="props.isFilter && isShowOptions">
        <UiInput
          v-model="filterQuery"
          :disabled="props.isDisabled"
          :placeholder="MESSAGES[props.lang].filter"
          isFocus
          data-test="ui-select-input-filter"
        />
      </div>

      <button
        v-if="props.isClearable && props.modelValue"
        @click="clearValue"
        type="button"
        :class="$style.clear"
        data-test="ui-select-clear"
      >
        Ã—
      </button>
    </div>

    <div
      v-if="isShowOptions"
      :class="[$style.options, { [$style.optionsUp]: isOpenUp }]"
      ref="optionsElement"
      data-test="ui-select-options"
    >
      <div v-if="optionsComputed.length > 0" ref="optionsInnerElement">
        <div
          v-for="(option, index) in optionsComputed"
          :key="getOptionKey(option, index)"
          @click="() => setOption(option)"
          @keydown.enter="() => setOption(option)"
          @keydown.space="() => setOption(option)"
          @mouseenter="() => setFocusedOptionIndex(index)"
          @keydown.up.prevent="() => setFocusedOptionIndex(index - 1)"
          @keydown.down.prevent="() => setFocusedOptionIndex(index + 1)"
          @keydown.esc="toggleOptions"
          :class="$style.option"
          tabindex="0"
          ref="optionElement"
          :data-current="isCurrentOption(option)"
          data-test="ui-select-option"
        >
          {{ getOptionTitle(option) }}
        </div>
      </div>

      <div v-else @click="hideOptions" :class="$style.option" tabindex="0" data-test="ui-select-no-results">
        {{ MESSAGES[props.lang].noResults }}
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { ref, computed } from 'vue';
import { onClickOutside } from '@vueuse/core';

import UiInput from '../UiInput/UiInput.vue';

import IconClosed from './icons/closed.svg?component';
import IconOpened from './icons/opened.svg?component';

import { TLocale } from '@/components/locales/types';
import { MESSAGES } from '@/components/locales';

interface IOption {
  _id?: string;
  title: string;
  title_en?: string;
}

interface IProps {
  modelValue?: string | number | IOption | null;
  options?: string[] | number[] | IOption[];
  isFilter?: boolean;
  isDisabled?: boolean;
  lang?: TLocale;
  isLocaleField?: boolean;
  isClearable?: boolean;
}

interface IEmit {
  'update:modelValue': [value: string | number | IOption | undefined];
  reachedBottom: [];
}

const props = withDefaults(defineProps<IProps>(), {
  modelValue: undefined,
  options: () => [],
  lang: 'ru',
});
const emit = defineEmits<IEmit>();

const filterQuery = ref('');
const isShowOptions = ref(false);
const isOpenUp = ref(false);

const containerElement = ref<HTMLElement>();
const optionsElement = ref<HTMLElement>();
const optionsInnerElement = ref<HTMLElement>();
const optionElement = ref<HTMLElement[]>([]);

const MIN_OPTIONS_HEIGHT = 240;

const isObject = computed(() => {
  const firstOption = props.options?.[0];

  return firstOption != undefined && typeof firstOption === 'object';
});

const isPrimitiveValue = (value: unknown): value is string | number => {
  return value !== null && (typeof value === 'string' || typeof value === 'number');
};

const optionsComputed = computed(() => {
  if (!props.options?.length) return [];

  const optionsArray: IOption[] = isObject.value
    ? [...(props.options as IOption[])]
    : (props.options as (string | number)[]).map((item) => ({ _id: String(item), title: String(item) }));

  return props.isFilter
    ? optionsArray.filter((option) => option.title.toLowerCase().includes(filterQuery.value.toLowerCase()))
    : optionsArray;
});

const displayValue = computed(() => {
  if (isPrimitiveValue(props.modelValue)) return props.modelValue;
  if (!props.modelValue) return '';

  const option = props.modelValue as IOption;

  return props.isLocaleField ? (option.title_en ?? '') : (option.title ?? '');
});

const getOptionKey = (option: IOption, index: number) => `${option._id || option.title}-${index}`;

const getOptionTitle = (option: IOption) => (props.isLocaleField ? option.title_en : option.title);

const isCurrentOption = (option: IOption) => {
  if (isPrimitiveValue(props.modelValue)) return props.modelValue === (option._id || option.title);
  if (props.modelValue) return props.modelValue._id === option._id;

  return false;
};

const setFocusedOptionIndex = (index: number) => {
  if (index < 0 || index >= optionsComputed.value.length || props.isFilter) return;
  optionElement.value[index]?.focus();
};

const clearValue = () => emit('update:modelValue', undefined);

const toggleOptions = () => {
  if (isShowOptions.value) {
    hideOptions();
  } else {
    showOptions();
  }
};

const checkOpenDirection = () => {
  if (!containerElement.value) return false;

  const dialogElement = containerElement.value.closest('dialog') as HTMLElement | null;

  if (!dialogElement) {
    const containerRect = containerElement.value.getBoundingClientRect();
    const spaceBelow = window.innerHeight - containerRect.bottom;
    const spaceAbove = containerRect.top;

    return spaceBelow < MIN_OPTIONS_HEIGHT && spaceAbove > spaceBelow;
  }

  const dialogRect = dialogElement.getBoundingClientRect();
  const containerRect = containerElement.value.getBoundingClientRect();

  const spaceBelowContainer = dialogRect.bottom - containerRect.bottom;
  const spaceAboveContainer = containerRect.top - dialogRect.top;

  return spaceBelowContainer < MIN_OPTIONS_HEIGHT && spaceAboveContainer > spaceBelowContainer;
};

function hideOptions() {
  optionsElement.value?.removeEventListener('scroll', checkScroll, true);
  filterQuery.value = '';
  isShowOptions.value = false;
  isOpenUp.value = false;
}

function showOptions() {
  if (props.isDisabled) return;

  isOpenUp.value = checkOpenDirection();
  isShowOptions.value = true;

  if (!props.isFilter) {
    setTimeout(() => {
      optionsElement.value?.scrollTo(0, 0);
      setFocusedOptionIndex(0);
      optionsElement.value?.addEventListener('scroll', checkScroll, true);
    }, 100);
  }
}

function setOption(option: IOption) {
  const value = isObject.value ? option : option._id || option.title;

  emit('update:modelValue', value);
  hideOptions();
}

function checkScroll() {
  if (!optionsElement.value || !optionsInnerElement.value) return;

  const scrollPosition =
    optionsElement.value.getBoundingClientRect().y -
    optionsInnerElement.value.getBoundingClientRect().y -
    optionsInnerElement.value.scrollHeight +
    optionsElement.value.getBoundingClientRect().height;

  if (scrollPosition > -100) emit('reachedBottom');
}

onClickOutside(containerElement, hideOptions);
</script>

<style module lang="scss">
.container {
  position: relative;

  &:disabled {
    color: var(--color-gray-dark-extra);
    background: var(--color-gray-light-extra);
    border-color: var(--color-transparent);
  }
}

.input {
  display: flex;
  gap: 8px;
  align-items: center;
}

.clear {
  width: 28px;
  height: 28px;
  font-size: 1rem;
  font-weight: 700;
  color: var(--color-white);
  cursor: pointer;
  background-color: var(--color-gray-dark);
  border: none;
  border-radius: 50%;

  &:hover {
    background-color: var(--color-gray-dark-extra);
  }
}

.options {
  position: absolute;
  z-index: 2;
  display: flex;
  flex-direction: column;
  min-width: 100%;
  max-height: 200px;
  margin-top: 8px;
  overflow-y: auto;
  background-color: var(--color-white);
  border: 1px solid var(--color-gray);
  border-radius: 16px;

  &.optionsUp {
    bottom: 100%;
    margin-top: 0;
    margin-bottom: 8px;
  }
}

.option {
  display: flex;
  flex-shrink: 0;
  align-items: center;
  height: 40px;
  padding: 0 16px;
  color: var(--color-black);
  cursor: pointer;
  outline: none;

  &:hover,
  &:focus {
    color: var(--color-white);
    background-color: var(--color-primary);
  }

  &[data-current='true'] {
    color: var(--color-black);
    background-color: var(--color-gray-light);

    &:hover,
    &:focus {
      background-color: var(--color-gray);
    }
  }
}

:global(.dark) {
  .options {
    background-color: var(--color-gray-dark);
    border: 1px solid var(--color-gray-dark-extra);

    [data-current='true'] {
      background-color: var(--color-gray-dark-extra);
    }
  }
}
</style>
