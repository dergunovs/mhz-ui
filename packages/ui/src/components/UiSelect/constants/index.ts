export const MODEL_VALUE = { _id: '1', title: 'Один', title_en: 'One' };

export const IS_FILTER = false;

export const OPTIONS = ['1', '2', '3', '4', '5', '6', '7', '8', '11', '111', '12'];

export const OPTIONS_OBJECTS = [
  { _id: '1', title: 'Один', title_en: 'One' },
  { _id: '2', title: 'Два', title_en: 'Two' },
  { _id: '3', title: 'Три', title_en: 'Three' },
];
