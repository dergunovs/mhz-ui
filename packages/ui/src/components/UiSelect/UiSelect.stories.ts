import { Meta, StoryObj } from '@storybook/vue3';
import { useArgs } from '@storybook/preview-api';

import { MODEL_VALUE, IS_FILTER, OPTIONS_OBJECTS } from './constants';

import { html } from '@/utils';
import { UiSelect } from '@/components';

const meta: Meta<typeof UiSelect> = {
  component: UiSelect,
  args: {
    modelValue: MODEL_VALUE,
    options: OPTIONS_OBJECTS,
    isFilter: IS_FILTER,
  },
  parameters: {
    docs: {
      description: {
        component: '',
      },
    },
  },
};

const argTypes = {
  modelValue: { description: 'string | { _id?: string; title: string }' },
  options: { description: 'string[] | { _id?: string; title: string }[]' },
};

type Story = StoryObj<typeof UiSelect>;

export default meta;

export const Primary: Story = {
  args: {
    modelValue: MODEL_VALUE,
  },

  render: (args, { updateArgs }) => ({
    components: { UiSelect },
    setup: () => ({ args, argTypes, updateArgs }),

    template: html` <UiSelect v-bind="args" @update:modelValue="update" />`,

    methods: {
      update(option: string) {
        updateArgs({ modelValue: option });
      },
    },
  }),

  decorators: [
    (story, context) => {
      const [args, updateArgs] = useArgs();

      return story({ ...context, updateArgs, args });
    },
  ],

  argTypes,
};

export const Filter: Story = {
  render: (args, { updateArgs }) => ({
    components: { UiSelect },
    setup: () => ({ args, argTypes, updateArgs }),

    template: html` <UiSelect v-bind="args" @update:modelValue="update" isFilter />`,

    methods: {
      update(option: string) {
        updateArgs({ modelValue: option });
      },
    },
  }),

  decorators: [
    (story, context) => {
      const [args, updateArgs] = useArgs();

      return story({ ...context, updateArgs, args });
    },
  ],

  argTypes,
};
