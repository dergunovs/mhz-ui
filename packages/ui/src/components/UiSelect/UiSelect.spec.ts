import { DefineComponent, nextTick } from 'vue';
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import UiSelect from './UiSelect.vue';
import { MODEL_VALUE, OPTIONS, IS_FILTER, OPTIONS_OBJECTS } from './constants';

import { wrapperFactory } from '@/test';

const selectInput = dataTest('ui-select-input');
const selectInputFilter = dataTest('ui-select-input-filter');
const selectOptions = dataTest('ui-select-options');
const selectOption = dataTest('ui-select-option');
const selectNoResults = dataTest('ui-select-no-results');
const selectClearButton = dataTest('ui-select-clear');

let wrapper: VueWrapper<InstanceType<typeof UiSelect>>;

beforeEach(() => {
  wrapper = wrapperFactory(UiSelect, { modelValue: MODEL_VALUE, options: OPTIONS, isFilter: IS_FILTER });
});

enableAutoUnmount(afterEach);

describe('UiSelect', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(UiSelect)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('toggles options on input click', async () => {
    expect(wrapper.find(selectOptions).exists()).toBe(false);

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.find(selectOptions).exists()).toBe(true);

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.find(selectOptions).exists()).toBe(false);
  });

  it('shows options', async () => {
    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.findAll(selectOption).length).toBe(OPTIONS.length);
    expect(wrapper.findAll(selectOption)[0].text()).toBe(OPTIONS[0]);
  });

  it('sets options', async () => {
    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    await wrapper.findAll(selectOption)[0].trigger('click');

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([OPTIONS[0]]);
  });

  it('shows object options title', async () => {
    await wrapper.setProps({ options: OPTIONS_OBJECTS });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.findAll(selectOption)[0].text()).toBe(OPTIONS_OBJECTS[0].title);
  });

  it('sets object options', async () => {
    await wrapper.setProps({ options: OPTIONS_OBJECTS });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    await wrapper.findAll(selectOption)[0].trigger('click');

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([OPTIONS_OBJECTS[0]]);
  });

  it('shows filter input in filter mode', async () => {
    await wrapper.setProps({ isFilter: true });

    expect(wrapper.find(selectInputFilter).exists()).toBe(false);

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.find(selectInputFilter).exists()).toBe(true);
  });

  it('shows computed results in filter mode', async () => {
    await wrapper.setProps({ isFilter: true });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.findAll(selectOption).length).toBe(OPTIONS.length);

    wrapper.findComponent<DefineComponent>(selectInputFilter).vm.$emit('update:modelValue', OPTIONS[1]);

    await nextTick();

    expect(wrapper.findAll(selectOption).length).toBe(
      OPTIONS.filter((option) => option.toLowerCase().includes(OPTIONS[1].toLowerCase())).length
    );
  });

  it('shows no results in filter mode', async () => {
    await wrapper.setProps({ isFilter: true });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.find(selectNoResults).exists()).toBe(false);

    wrapper.findComponent<DefineComponent>(selectInputFilter).vm.$emit('update:modelValue', '100% not in options');

    await nextTick();

    expect(wrapper.findAll(selectOption).length).toBe(0);
    expect(wrapper.find(selectNoResults).exists()).toBe(true);

    await wrapper.find(selectNoResults).trigger('click');

    expect(wrapper.find(selectOptions).exists()).toBe(false);
  });

  it('clears value when clear button is clicked', async () => {
    expect(wrapper.find(selectClearButton).exists()).toBe(false);

    await wrapper.setProps({ isClearable: true });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.find(selectClearButton).exists()).toBe(true);

    await wrapper.find(selectClearButton).trigger('click');

    expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([undefined]);
  });

  it('disables component when isDisabled prop is true', async () => {
    expect(wrapper.find(selectOptions).exists()).toBe(false);

    await wrapper.setProps({ isDisabled: true });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.find(selectOptions).exists()).toBe(false);
  });

  it('shows placeholder text in different languages', async () => {
    await wrapper.setProps({ lang: 'en' });

    expect(wrapper.findComponent<DefineComponent>(selectInput).attributes('placeholder')).toBe('Choose variant');

    await wrapper.setProps({ lang: 'ru' });

    expect(wrapper.findComponent<DefineComponent>(selectInput).attributes('placeholder')).toBe('Выбрать');
  });

  it('shows filter placeholder text in different languages', async () => {
    await wrapper.setProps({ isFilter: true, lang: 'en' });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    const filterComponent = wrapper.findComponent(selectInputFilter) as VueWrapper;

    expect(filterComponent.attributes('placeholder')).toBe('Filter Variants');
  });

  it('shows no results text in different languages', async () => {
    await wrapper.setProps({ isFilter: true, lang: 'en' });

    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    wrapper.findComponent<DefineComponent>(selectInputFilter).vm.$emit('update:modelValue', 'nonexistent');

    await nextTick();

    expect(wrapper.find(selectNoResults).text()).toBe('No results');
  });

  it('handles keyboard navigation in options', async () => {
    wrapper.findComponent<DefineComponent>(selectInput).vm.$emit('toggle');

    await nextTick();

    expect(wrapper.findAll(selectOption)[0].attributes('tabindex')).toBe('0');
  });
});
