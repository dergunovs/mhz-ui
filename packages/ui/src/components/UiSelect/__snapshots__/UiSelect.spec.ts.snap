// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UiSelect > matches snapshot 1`] = `
"<div class="_container_cf7a81">
  <div class="_input_cf7a81">
    <ui-input-stub modelvalue="" type="text" isdisabled="false" mode="select" appendicon="[object Object]" isfocus="false" iscenter="false" iscopy="false" ispassword="false" disabled="false" placeholder="Choose variant" data-test="ui-select-input"></ui-input-stub>
    <!--v-if-->
    <!--v-if-->
  </div>
  <!--v-if-->
</div>"
`;
